### ISSUE TITLE
Support network topologies where multiple services send/receive on a single channel

### ISSUE THREAD

**Issue by shadaj** [hydro-project/hydroflow#398]
Right now, we only support a single service's output connected to a single service's input port, but this is too restrictive. I propose an API that allows service ports to be grouped together some some identifiers to allow multiple recipients / senders on a single port.

```python
FanInPort({
    0: a.ports.foo,
    1: b.ports.bar
}).send_to(c.ports.baz)
```

would result in `baz` receiving values of the form `(0, data from foo)`, `(1, data from bar)`. These IDs do not necessarily need to be sent over the network, if we have a guarantee that, for example, the source IP address uniquely identify the service. We can have a similar API for `FanOutPort`, where the HF program is responsible for tagging data with which service they want to send data to. This most closely corresponds to the Dedalus model of networking.

**Comment by jhellerstein** [comment]
Isnt this already supported via demux? How is a "port" abstraction different?Sent from a telephoneOn Feb 20, 2023, at 11:10 AM, Shadaj Laddad ***@***.***> wrote:﻿
Right now, we only support a single service's output connected to a single service's input port, but this is too restrictive. I propose an API that allows service ports to be grouped together some some identifiers to allow multiple recipients / senders on a single port.
FanInPort({
    0: a.ports.foo,
    1: b.ports.bar
}).send_to(c.ports.baz)
would result in baz receiving values of the form (0, data from foo), (1, data from bar). These IDs do not necessarily need to be sent over the network, if we have a guarantee that, for example, the source IP address uniquely identify the service. We can have a similar API for FanOutPort, where the HF program is responsible for tagging data with which service they want to send data to. This most closely corresponds to the Dedalus model of networking.

**Comment by shadaj** [comment]
Oh, this is not meant as a bug on `hydroflow`, more on the CLI to handle wiring things up. I expect that we will use `demux` at runtime.

**Comment by shadaj** [comment]
Closed in #445
